apollo-federation/src/sources/connect/json_selection/selection_trie.rs

```diff
@@ -0,0 +1,390 @@
+use std::fmt::Display;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::ops::Range;
+
+use apollo_compiler::collections::IndexMap;
+use apollo_compiler::collections::IndexSet;
+
+use super::Key;
+use super::NamedSelection;
+use super::PathList;
+use super::Ranged;
+use super::SubSelection;
+use super::helpers::quote_if_necessary;
+use super::location::WithRange;
+
+impl PathList {
+    pub(crate) fn compute_selection_trie(&self) -> SelectionTrie {
+        let mut trie = SelectionTrie::new();
+        trie.add_path_list(self);
+        trie
+    }
+}
+
+type Ref<T> = std::sync::Arc<T>;
+
+#[derive(Debug, Eq, Clone)]
+pub(crate) struct SelectionTrie {
+    /// The top-level sub-selections of this [`SelectionTrie`].
+    selections: IndexMap<String, Ref<SelectionTrie>>,
+
+    /// Whether the path terminating at this [`SelectionTrie`] node was
+    /// explicitly added to the trie.
+    used: bool,
+
+    /// Collected as metadata but ignored by [`PartialEq`] and [`Hash`].
+    key_ranges: IndexMap<String, IndexSet<Range<usize>>>,
+}
+
+impl Display for SelectionTrie {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut need_space = false;
+
+        for (key, sub) in self.selections.iter() {
+            if need_space {
+                write!(f, " ")?;
+            }
+
+            if sub.is_empty() {
+                if sub.is_used() {
+                    write!(f, "{}", quote_if_necessary(key))?;
+                    need_space = true;
+                }
+            } else {
+                write!(f, "{} {{ {} }}", quote_if_necessary(key), sub)?;
+                need_space = true;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl PartialEq for SelectionTrie {
+    fn eq(&self, other: &Self) -> bool {
+        self.used == other.used && self.selections == other.selections
+    }
+}
+
+impl Hash for SelectionTrie {
```

Comment:
Can't you do the following here:
```rust

impl Hash for SelectionTrie {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.selections.hash(state);
    }
}
```

Should work

Reply:
That would work if `IndexMap` implemented `Hash`! 😞  See https://github.com/indexmap-rs/indexmap/issues/67 if you don't mind being a little disappointed.

I personally think `IndexMap` should support an implementation of `Hash` out of the box. Even though `IndexMap` preserves insertion order, `IndexMap`'s implementation of `PartialEq` is order-insensitive, so I think it also can/should implement `Hash` with an order-insensitive combination operator (note: XOR/`^` may not be the best choice here). Until the maintainers come around and implement `Hash` natively, my workaround seems necessary.

Now, we don't have to use `IndexMap` here, especially if we end up normalizing tries by sorting their keys (or something). But if we're using `IndexMap` then we need to hash it manually, I think.